File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode temp=head;
        if(temp==null){
            return null;

        }
        ListNode prev=null;
        while(temp!=null && temp.next!=null){
            if(temp.val==temp.next.val){
                while(temp!=null && temp.next!=null && temp.val==temp.next.val){
                    temp=temp.next;
                }
                if(prev==null){
                    head=temp.next;
                    
                   
                    
                }
                else{
                    prev.next=temp.next;
                    
                }
                temp=temp.next;
            }
            else{
                prev=temp;
                temp=temp.next;
            }
        }
        return head;
        
    }
}

File: 0082-remove-duplicates-from-sorted-list-ii/0082-remove-duplicates-from-sorted-list-ii.c
struct ListNode* deleteDuplicates(struct ListNode* head) {
    if (head == NULL || head->next == NULL)
        return head;

    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    dummy->val = -1;  // dummy value
    dummy->next = head;

    struct ListNode* prev = dummy;
    struct ListNode* curr = head;

    while (curr != NULL) {
        bool duplicate = false;

        while (curr->next != NULL && curr->val == curr->next->val) {
            duplicate = true;
            curr = curr->next;
        }

        if (duplicate) {
            prev->next = curr->next;
        } else {
            prev = prev->next;
        }

        curr = curr->next;
    }

    struct ListNode* result = dummy->next;
    free(dummy);

    return result;
}
